bundles/org.openhab.binding.hive/src/main/java/org/openhab/binding/hive/internal/handler/HiveBridgeHandler.java

```diff
@@ -0,0 +1,394 @@
+                if (e.getMessage().contains("WWW-Authenticate header")) {
+                    getToken();
+                } else {
+                    logger.warn("Failed to get new token from Hive API: {}", e.getMessage());
+                }
+                return;
+            }
+            int statusCode = response.getStatus();
+
+            if (statusCode == HttpStatus.UNAUTHORIZED_401) {
+                // Token expired, get a new one and try again
+                getToken();
+                statusCode = response.getStatus();
+            }
+
+            if (statusCode != HttpStatus.OK_200) {
+                String statusLine = response.getStatus() + " " + response.getReason();
+                logger.warn("Error while reading from Hive API: {}", statusLine);
+                return;
+            }
+
+            HiveNodes o = gson.fromJson(response.getContentAsString(), HiveNodes.class);
+
+            if (o.nodes.size() > 0) {
+                // Loop through the nodes and add thermostats but only if they are showing a temperature as the api
+                // reports lots of thermostats for some reason
+                HiveNode thermostat = null;
+                HiveNode systemInfo = null;
+                for (HiveNode node : o.nodes) {
+                    if (node.attributes != null && node.attributes.nodeType != null
+                            && node.attributes.nodeType.reportedValue.equals(THERMOSTAT_NODE_TYPE)
+                            && node.attributes.temperature != null) {
+                        // If this a thermostat, and it has a current temperature, add it to discovery results
+                        thermostat = node;
+                    }
+                    if (node.attributes != null && node.attributes.batteryLevel != null) {
+                        // If this a thermostat, and it has a current temperature, add it to discovery results
+                        systemInfo = node;
+                    }
+                }
+                if (thermostat != null) {
+                    if (systemInfo != null) {
+                        thermostat.linkedNode = systemInfo.id;
+                        thermostat.firmwareVersion = systemInfo.attributes.softwareVersion.displayValue;
+                        thermostat.model = systemInfo.attributes.model.displayValue;
+                        thermostat.macAddress = systemInfo.attributes.macAddress.displayValue;
+                    }
+                    if (hiveDiscoveryService != null) {
+                        hiveDiscoveryService.addDevice(thermostat);
+                    }
+                }
+            }
+        }
+    }
+
+    public HiveAttributes getThermostatReading(Thing thing) {
+        HiveAttributes reading = new HiveAttributes();
+
+        if (online) {
+            ContentResponse response;
+            int statusCode = 0;
+            String responseString = "";
+            HiveNodes o = null;
```

Comment:
I prefer to keep the scope of local variables as limited as possible, so declaring it where it is used would be great for that. Also, the method should ideally be not as big as it is now.